sass/sass.bzl

```diff
@@ -11,219 +11,170 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
+"Compile Sass files to CSS"
 
-SASS_FILETYPES = FileType([
-    ".sass",
-    ".scss",
-])
+_FILETYPES = [".sass", ".scss", ".svg", ".png", ".gif"]
 
-def collect_transitive_sources(ctx):
-    source_files = depset(order="postorder")
-    for dep in ctx.attr.deps:
-        source_files += dep.transitive_sass_files
-    return source_files
+# Documentation for switching which compiler is used
+_COMPILER_ATTR_DOC = """Choose which Sass compiler binary to use.
+
+By default, we use the JavaScript-transpiled version of the
+dart-sass library, based on https://github.com/sass/dart-sass.
+This is the canonical compiler under active development by the Sass team.
+This compiler is convenient for frontend developers since it's released
+as JavaScript and can run natively in NodeJS without being locally built.
+However, it is the slowest option. If you have a substantial Sass
+codebase, consider the two other options for this attribute:
+
+1. `compiler = "@sassc"` uses the libsass compiler written in C++.
+   As of 2018, this is the most commonly used compiler.
+   It requires your Bazel setup has a working C++ compilation toolchain.
+
+   NOTE: future releases of rules_sass may remove this option, if it
+   becomes obsolete or if the maintenance burden is too high.
+
+2. Dart Sass runs the new compiler implementation natively in the Dart
+   VM. This option requires a change to the Dart Bazel rules which is
+   not yet available as of May 2018.
+"""
+
+SassInfo = provider(
+    doc = "Collects files from sass_library for use in downstream sass_binary",
+    fields = {
+        "transitive_sources": "Sass sources for this target and its dependencies",
+    })
+
+def _collect_transitive_sources(srcs, deps):
+  "Sass compilation requires all transitive .sass source files"
+  return depset(
+      srcs,
+      transitive=[dep[SassInfo].transitive_sources for dep in deps],
+      # Provide .sass sources from dependencies first
+      order="postorder")
 
 def _sass_library_impl(ctx):
-    transitive_sources = collect_transitive_sources(ctx)
-    transitive_sources += SASS_FILETYPES.filter(ctx.files.srcs)
-    return struct(
-        files = depset(),
-        transitive_sass_files = transitive_sources)
+  """sass_library collects all transitive sources for given srcs and deps.
+  It doesn't execute any actions."""
+  transitive_sources = _collect_transitive_sources(
+      ctx.files.srcs, ctx.attr.deps)
+  return [SassInfo(transitive_sources=transitive_sources)]
+
+def _run_sass(ctx, input, css_output, map_output):
+  """run_sass performs an action to compile a single Sass file into CSS."""
+  # The Sass CLI expects inputs like
+  # sass <flags> <input_filename> <output_filename>
+  args = ctx.actions.args()
+
+  # Flags (see https://github.com/sass/node-sass#options)
+  # Note, the command line is compatible with Dart sass.
```

Comment:
There's some overlap between Dart Sass's and Node Sass's command-line options, but they aren't identical and we don't intend to retain compatibility on principle.

Reply:
Is the JS output version of the Dart compiler API-equivalent with node-sass? I had remembered you mentioning that it was supposed to be a drop-in replacement. 

Is there a better place to point for the Dart compiler flags?

Reply:
> Is the JS output version of the Dart compiler API-equivalent with node-sass? I had remembered you mentioning that it was supposed to be a drop-in replacement.

It's JS-API-equivalent, but not CLI-equivalent. We expect most automation to go through the JS API, and for the CLI to be generally human-facing; Bazel is unusual in that regard.

> Is there a better place to point for the Dart compiler flags?

Unfortunately there isn't (again because we tend to expect the CLI to be human-facing). I guess [the source file where they're defined](https://github.com/sass/dart-sass/blob/master/lib/src/executable/options.dart)?

Reply:
Removed the mention of API equivalence and pointed to the source file.